### ISSUE TITLE
Finetune of SPHINX Mixtral MoE - Questions

### ISSUE THREAD

**Issue by babla9** [Alpha-VLLM/LLaMA2-Accessory#156]
Thanks for the great work here! 

I'm following along on the guide here to fine-tune the mixtral MoE version of Sphinx. https://github.com/Alpha-VLLM/LLaMA2-Accessory/tree/main/SPHINX#finetune-sphinx 

I have a few questions: 

1. What are the right llama config values for fine-tuning Mixtral MoE? If I download the right mistral checkpoint from [here](https://huggingface.co/Alpha-VLLM/LLaMA2-Accessory/tree/main/finetune/mm/SPHINX/SPHINX-MoE-1k) and specify 'mistral_sparse_ens'  as llama_type in the [fine tuning config](https://github.com/Alpha-VLLM/LLaMA2-Accessory/tree/main/SPHINX#finetuning), would that be all that's required? 
2. Do the provided image files in the dataset have to be scaled to certain dimensions? 
3. What is the fine-tuning time for ~10k samples on an 8x A100 cluster? 

Thank you!

**Comment by ChrisLiu6** [comment]
> If I download the right mistral checkpoint from [here](https://huggingface.co/Alpha-VLLM/LLaMA2-Accessory/tree/main/finetune/mm/SPHINX/SPHINX-MoE-1k) and specify 'mistral_sparse_ens' as llama_type in the [fine tuning config](https://github.com/Alpha-VLLM/LLaMA2-Accessory/tree/main/SPHINX#finetuning), would that be all that's required?

Yes, please also make sure to 1. use the tokenizer we provide along with checkpoints and 2. configure model parallel size (we use 8) and gradient checkpointing (we turn it on) properly.

>  Do the provided image files in the dataset have to be scaled to certain dimensions?
You don't need to do it manually. The transformation is already implemented in the training pipeline

> 8 x A100 may not be enough to finetune Mixtral 8x7B. On our cluster, with 32 GPUs, 8-way model parallel, and gradient checkpionting, the training speed is approximated 2s / minibatch

**Comment by babla9** [comment]
@ChrisLiu6 thank you for your response, this is very helpful! 

> 8 x A100 may not be enough to finetune Mixtral 8x7B. On our cluster, with 32 GPUs, 8-way model parallel, and gradient checkpionting, the training speed is approximated 2s / minibatch

Some follow up questions on this. 
- Do you mean that it will take a long time to run or that the 8 x a100 cluster will run out of memory? 
- How big is a mini batch typically? If it scales linearly, assuming mini batch is 1 sample, at 8s/minibatch * 10,000 batches /3600 = ~22 hours? 
- Can I enable lora here to help?

**Comment by ChrisLiu6** [comment]
> @ChrisLiu6 thank you for your response, this is very helpful!
> 
> > 8 x A100 may not be enough to finetune Mixtral 8x7B. On our cluster, with 32 GPUs, 8-way model parallel, and gradient checkpionting, the training speed is approximated 2s / minibatch
> 
> Some follow up questions on this.
> 
> * Do you mean that it will take a long time for the fine-tuning to run or that the 8 x a100 cluster will run out of memory?
> * How big is a mini batch typically? Worst case, if it scales linearly, assuming mini batch is just 1 sample, at 8s/minibatch * 10,000 batches /3600 = ~22 hours using an 8xA100?
> * Can I enable lora here to help?
> * As a last resort, would it make sense to use llama2 and Sphinx2 (instead of MoE)?

1. It will run out of memory. Specifically, with 8-way model parallel (i.e. Megetron-style tensor parallel), the model is split into 8 parts, each part approximately takes 10G memory in fp16/bf16 format. During training, each GPU takes one part. which means the memory cost will be at least 10G (model parameter, fp16/bf16) + 20G (gradient, fp32) + 40G (20G for Adam beta1 and 20G for Adam beta2) without considering any intermediate activations produced during forward/backward. When using 16 GPUs, there will be two GPUs to take charge of the same part, then the load for 20G gradient and 40G optimizer states will be halved on each GPU through zero2. Overall, to full finetune mixtral, would suggest at least 16 GPUs.
2. Sorry, I should have made it clearer. We use mini batch size 4. Due to 8-way model parallel, when 32 GPUs are used, there are actually 4 data parallel groups, so the effective batch size (without considering gradient accumulation) is 16. On this condition, the speed is about 2s/minibatch
3. Yes, LoRA could help a lot. But unfortunately, we have yet to implement LoRA+Mixtral *sparse*, as adding LoRA to the sparse-implemented MoE layer has some additional complexity (but theoretically it should be 100% feasible). We have only implemented the LoRA + Mixtral *base* implementation.
4.  Of course you can the LLaMA2-version SPHINX. In fact, its performance is even better than SPHINX-MoE on several benchmarks, and it is also used as the backend for our [web-demo](http://imagebind-llm.opengvlab.com/)

**Comment by babla9** [comment]
@ChrisLiu6 thank you so much for the detailed explanation, this clears up many questions I had!

> Of course you can the LLaMA2-version SPHINX. In fact, its performance is even better than SPHINX-MoE on several benchmarks, and it is also used as the backend for our [web-demo](http://imagebind-llm.opengvlab.com/) 

I will use SPHINX-v2-1k for now given my limited compute. 
- Do you know what the fine-tuning speed is for this model on 8xA100? 
- Am I right in understanding that for this model, max_words can be up to 1000?

Thanks!

**Comment by kumagai6** [comment]
Where is zero2 implemented in the code?

**Comment by ChrisLiu6** [comment]
> Where is zero2 implemented in the code?

PyTorch FSDP is roughly an equivalent reproduction of deepspeed Zero
